src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md

```diff
@@ -0,0 +1,549 @@
+
+如果你调用 `Person('Fred')` 时 **没有** 加 `new`，其中的 `this` 会指向某个全局且无用的东西（比如，`window` 或者 `undefined`），因此我们的代码会崩溃，或者做一些像设置 `window.name` 之类的傻事。
+
+通过在调用前增加 `new`，我们说：“嘿 JavaScript，我知道 `Person` 只是个函数，但让我们假装它是个构造函数吧。**创建一个 `{}` 对象并把 `Person` 中的 `this` 指向那个对象，以便我可以通过类似 `this.name` 的形式去设置一些东西，然后把这个对象返回给我。**”
+
+这就是 `new` 操作符所做的事。
+
+```jsx
+var fred = new Person('Fred'); // 和 `Person` 中的 `this` 等效的对象
+```
+
+`new` 操作符同时也把我们放在 `Person.prototype` 上的东西放到了 `fred` 对象上：
+
+```jsx
+function Person(name) {
+  this.name = name;
+}
+Person.prototype.sayHi = function() {  alert('Hi, I am ' + this.name);}
+var fred = new Person('Fred');
+fred.sayHi();
+```
+
+这就是在 JavaScript 直接支持类之前，人们模拟类的方式。
+
+* * *
+
+`new` 在 JavaScript 中已经存在了好久了，然而类还只是最近的事，它的出现让我们能够重构我们前面的代码以使它更符合我们的本意：
+
+```jsx
+class Person {
+  constructor(name) {
+    this.name = name;
+  }
+  sayHi() {
+    alert('Hi, I am ' + this.name);
+  }
+}
+
+let fred = new Person('Fred');
+fred.sayHi();
+```
+
+***捕捉开发者的本意*** 是语言和 API 设计中非常重要的一点。
+
+如果你写了一个函数，JavaScript 没办法判断它应该像 `alert()` 一样被调用，还是应该被视作像 `new Person()` 一样的构造函数。忘记给像 `Person` 这样的函数指定 `new` 会导致令人费解的行为。
+
+**类语法允许我们说：“这不仅仅是个函数 —— 这是个类并且它有构造函数”。** 如果你在调用它时忘了加 `new`，JavaScript 会报错：
+
+```jsx
+let fred = new Person('Fred');
+// ✅  如果 Person 是个函数：有效
+// ✅  如果 Person 是个类：依然有效
+
+let george = Person('George'); // 我们忘记使用 `new`
+// 😳 如果 Person 是个长得像构造函数的方法：令人困惑的行为
+// 🔴 如果 Person 是个类：立即失败
+```
+
+这可以帮助我们在早期捕捉错误，而不会遇到类似 `this.name` 被当成 `window.name` 对待而不是 `george.name` 的隐晦错误。
+
+然而，这意味着 React 需要在调用所有类之前加上 `new`，而不能把它直接当做一个常规的函数去调用，因为 JavaScript 会把它当做一个错误对待！
+
+```jsx
+class Counter extends React.Component {
+  render() {
+    return <p>Hello</p>;
+  }
+}
+
+// 🔴 React 不能简单这么做：
+const instance = Counter(props);
+```
+
+这意味着麻烦。
+
+* * *
+
+在我们看到 React 如何处理这个问题之前，很重要的一点就是要记得大部分 React 的用户会使用 Babel 等编译器来编译类等现代化的特性以便能在老旧的浏览器上运行。因此我们需要在我们的设计中考虑编译器。
+
+在 Babel 的早期版本中，类不加 `new` 也可以被调用。但这个问题已经被修复了 —— 通过生成额外的代码的方式。
+
+```jsx
+function Person(name) {
+  // 稍微简化了一下 Babel 的输出：
+  if (!(this instanceof Person)) {
+    throw new TypeError("Cannot call a class as a function");
+  }
+  // Our code:
+  this.name = name;
+}
+
+new Person('Fred'); // ✅ OK
+Person('George');   // 🔴 无法把类当做函数来调用
+```
+
+你或许已经在你构建出来的包中见过类似的代码，这就是那些 `_classCallCheck` 函数做的事。（你可以通过启用“loose mode”来关闭检查以减小构建包的尺寸，但这或许会使你最终转向真正的原生类时变得复杂）
+
+* * *
+
+至此，你应该已经大致理解了调用时加不加 `new` 的差别：
+
+|            | `new Person()`               | `Person()`                          |
+| ---------- | ---------------------------- | ----------------------------------- |
+| `class`    | ✅ `this` 是一个 `Person` 实例 | 🔴 `TypeError`                      |
+| `function` | ✅ `this` 是一个 `Person` 实例 | 😳 `this` 是 `window` 或 `undefined` |
+
+这就是 React 正确调用你的组件很重要的原因。 **如果你的组件被定义为一个类，React 需要使用 `new` 来调用它**
+
+所以 React 能检查出某样东西是否是类吗？
+
+没那么容易！即便我们能够 [在 JavaScript 中区分类和函数](https://stackoverflow.com/questions/29093396/how-do-you-check-the-difference-between-an-ecmascript-6-class-and-function)，面对被 Babel 等工具处理过的类这还是没用。对浏览器而言，它们只是不同的函数。这是 React 的不幸。
```

Comment:
```suggestion
没那么容易！即便我们能够[在 JavaScript 中区分类和函数](https://stackoverflow.com/questions/29093396/how-do-you-check-the-difference-between-an-ecmascript-6-class-and-function)，面对被 Babel 等工具处理过的类这还是没用。对浏览器而言，它们只是不同的函数。这是 React 的不幸。
```

Reply:
I think that a [space between links](https://github.com/mzlogin/chinese-copywriting-guidelines#%E9%93%BE%E6%8E%A5%E4%B9%8B%E9%97%B4%E5%A2%9E%E5%8A%A0%E7%A9%BA%E6%A0%BC)

Reply:
@tonghuashuo Ah, so I checked and looks like you can indeed choose between adding a space or not. It says it's controversial. In this case, you can skip my suggestions on space before links then.